bdk_core_example/src/esplora.rs

```diff
@@ -0,0 +1,166 @@
+            last_valid = Some(tip);
+            break;
+        } else {
+            invalidate = Some(tip);
+        }
+    }
+
+    let new_tip = BlockId {
+        height: client.get_height()?,
+        hash: client.get_tip_hash()?,
+    };
+
+    loop {
+        let handles: Vec<JoinHandle<Result<(_, Vec<_>), anyhow::Error>>> = (0..parallel_requests)
+            .filter_map(|_| {
+                let (index, script) = scripts.next()?;
+                let client = client.clone();
+                Some(std::thread::spawn(move || {
+                    let mut related_txs: Vec<Tx> = client.scripthash_txs(&script, None)?;
+
+                    let n_confirmed = related_txs.iter().filter(|tx| tx.status.confirmed).count();
+                    // esplora pages on 25 confirmed transactions. If there's 25 or more we
+                    // keep requesting to see if there's more.
+                    if n_confirmed >= 25 {
+                        loop {
+                            let new_related_txs: Vec<Tx> = client
+                                .scripthash_txs(&script, Some(related_txs.last().unwrap().txid))?;
+                            let n = new_related_txs.len();
+                            related_txs.extend(new_related_txs);
+                            // we've reached the end
+                            if n < 25 {
+                                break;
+                            }
+                        }
+                    }
+
+                    Ok((index, related_txs))
+                }))
+            })
+            .collect::<Vec<_>>();
+
+        let n_handles = handles.len();
+
+        for handle in handles {
+            let (index, related_txs) = handle.join().unwrap()?; // TODO: don't unwrap
+            if related_txs.is_empty() {
+                empty_scripts += 1;
+            } else {
+                last_active_index = Some(index);
+                empty_scripts = 0;
+            }
+            for tx in related_txs {
+                transactions.push((tx.to_tx(), tx.status.block_time))
+            }
+        }
+
+        if n_handles == 0 || empty_scripts >= stop_gap {
+            break;
+        }
+    }
+
+    if client.get_tip_hash()? != new_tip.hash {
+        return Err(anyhow!("Tip Moved during update"));
+    }
+
+    transactions.iter().for_each(|(tx, _)| {
+        tx_graph.insert_tx(tx);
+    });
+
+    let update = Update {
+        txids: transactions
+            .iter()
+            .map(|(tx, conf)| {
+                (
+                    tx.txid(),
+                    conf.map_or(TxHeight::Unconfirmed, |ht| {
+                        bdk_core::TxHeight::Confirmed(ht as u32)
+                    }),
+                )
+            })
+            .collect(),
+        last_valid,
+        invalidate,
+        new_tip,
+    };
+
+    Ok((last_active_index, update))
+}
+
+pub fn perform_esplora_sync<K: Debug + Clone + Ord>(
```

Comment:
This should be in the same file as `fn main()` (or better yet just in `fn main()`) for the esplora example. The logic in here is the thing you are trying to demonstrate to people so it should be the first thing they see when they open up the example.